providers/implementations/keymgmt/hybrid_kmgmt.c

```diff
@@ -0,0 +1,668 @@
+                                            unsigned int num_algs,
+                                            const char **names,
+                                            uint32_t *key_lengths)
+{
+    OSSL_LIB_CTX *libctx = PROV_LIBCTX_OF(provctx);
+    HYBRID_PKEY_CTX *ctx = NULL;
+    unsigned int i;
+
+    if (!ossl_prov_is_running())
+        return NULL;
+
+    if ((ctx = OPENSSL_zalloc(sizeof(*ctx))) != NULL) {
+        ctx->libctx = libctx;
+        ctx->propq = NULL;
+        ctx->num_algs = num_algs;
+        ctx->names = names;
+        ctx->key_lengths = key_lengths;
+        for (i = 0; i < ctx->num_algs; i++) {
+            ctx->ctxs[i] = EVP_PKEY_CTX_new_from_name(libctx, ctx->names[i],
+                                                      ctx->propq);
+            if (ctx->ctxs[i] == NULL
+                    || !EVP_PKEY_keygen_init(ctx->ctxs[i]))
+                goto err;
+        }
+        if (!ossl_hybrid_kmgmt_gen_set_params(ctx, params))
+            goto err;
+    }
+    return ctx;
+ err:
+    ossl_hybrid_pkey_ctx_free(ctx);
+    return NULL;
+}
+
+int ossl_hybrid_get_ctx_params(HYBRID_PKEY_CTX *ctx, OSSL_PARAM params[])
+{
+    OSSL_PARAM prms[2] = { OSSL_PARAM_END, OSSL_PARAM_END };
+    unsigned char *q;
+    OSSL_PARAM *p;
+    uint32_t t;
+    unsigned int i;
+    int type;
+
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    for (p = params; params->key != NULL; params++) {
+        type = ossl_param_find_pidx(p->key);
+        switch (type) {
+        case PIDX_PKEY_PARAM_SECURITY_BITS:
+            t = maximum_ctx_uint32_params(ctx->num_algs, ctx->ctxs,
+                                          OSSL_PKEY_PARAM_SECURITY_BITS);
+            if (t == 0 || !OSSL_PARAM_set_uint32(p, t))
+                return 0;
+            break;
+
+        case PIDX_PKEY_PARAM_BITS:
+            t = accumulate_ctx_uint32_params(ctx->num_algs, ctx->ctxs,
+                                             OSSL_PKEY_PARAM_BITS, NULL);
+            if (t == 0 || !OSSL_PARAM_set_uint32(p, t))
+                return 0;
+            break;
+
+        case PIDX_PKEY_PARAM_MAX_SIZE:
+            t = ossl_hybrid_get_key_length_from_ctx(ctx);
+            if (t == 0 || !OSSL_PARAM_set_uint32(p, t))
+                return 0;
+            break;
+
+        case PIDX_PKEY_PARAM_ENCODED_PUBLIC_KEY:
+            t = ossl_hybrid_get_key_length_from_ctx(ctx);
+            if (t > p->data_size) {
+                p->data_size = t;
+                ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH,
+                               "length must be %u bytes, have %u bytes",
+                               t, p->data_size);
+                return 0;
+            }
+            p->data_size = t;
+            q = p->data;
+            *prms = *p;
+            for (i = 0; i < ctx->num_algs; i++) {
+                prms->data = q;
+                prms->data_size = ctx->key_lengths[i];
+                q += ctx->key_lengths[i];
```

Comment:
Should we have an assert somewhere here to make sure we haven't accidentally run off the end of `p->data`

Reply:
The check on line 494 will forbid this.
```c
            t = ossl_hybrid_get_key_length_from_ctx(ctx);
            if (t > p->data_size) {
                ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH,
                               "length must be %u bytes, have %u bytes",
                               t, p->data_size);
                p->data_size = t;
                return 0;
            }
```
I guess an assert isn't overly harmful.